Testes Travis CI and included Unit tests

jmolmo opened the pull request:
I have improve the CI execution duration caching the dependencies of pip packages.

Executing the unit tests i have seen that when the inventory file is created the AnsibleInventary" fd attribute wasn't properly initialized. 
I have made several changes, although i think that this can be improved ( probably we should use a database to keep the inventory and avoid concurrency problems, this will be even more healthy if we are running the tool using several instances in several servers)
Minor changes in the unit test ( basically removed  obsolete methods)

I have include the execution of the unit test in the CI Travis config file.

And i have build succesfully with Travis:
https://travis-ci.org/jmolmo/ansible-runner-service/builds/427189758

pcuzner commented:
database? runner expects to use inventory/hosts or have a host list passed, so I don't think introducing a database is viable.

pcuzner commented on runner_service/inventory.py:
```diff
@@ -91,37 +99,52 @@ def load(self):
                     inv.write(yaml.dump(AnsibleInventory.inventory_seed,
-                                        default_flow_style=False))
+                                            default_flow_style=False))
+            except FileExistsError:
```
FileExistsError....but you're inside a clause that says the file doesn't exists?

jmolmo replied:
But in a system with concurrency ... other instance of the service can be doing the same at the same time .... is a race condition

pcuzner replied:
Agree - it wouldn't be another service, but it could feasibly be another request

pcuzner commented on runner_service/inventory.py:
```diff
@@ -91,37 +99,52 @@ def load(self):
         if not raw:
-            self.inventory = None
+            # If the inventory is empty for some extrange reason
+            self.inventory = AnsibleInventory.inventory_seed
```
I don't think we should do this. If we can't read the inventory - we can't just seed it - we need to ensure the file is consistent with what we're working against. If we can't read it - we flag it and abort. What if the inventory did contain 'stuff' but the 'raw' was empty - we'd lose what was there!

pcuzner replied:
I guess given the logic I'm saying that I can't think of a scenario where raw would be empty - if it is I'd rather return None, so the inventory load will end up throwing an error to the caller.

jmolmo replied:
Ok. i'll change :-)

pcuzner commented on runner_service/inventory.py:
```diff
@@ -91,37 +99,52 @@ def load(self):
 
     def save(self):
+        # Get the only when fd when we are going to save
+        self.fd = open(self.filename, 'w')
```
why do we need to do this here? When the inventory is loaded and the exclusive lock is requested, the fs is locked until we exit/save - so not sure what the scenario is that this addresses?

jmolmo replied:
This is the behaviour when you cretate the inventory objects with: AnsibleInventory(excl=True)
If you do not set especifically excl=True, the fd attribute of the object is not initialized. and all the calls to the "save" method return error because the "seek" method can not be used in a "none" object.

I not sure about what do you wanted exactly...maybe to provide a "read" only inventory when the "excl"  parameter is not set to True? if it is the aim then the "save" method must be disabled  when excl=False.

I made this modification because in the unit tests the inventory is created with "Excl=false" and write operations are executed ( in fact all  the "write" test failed because fd not initialized).

Tell me what is what do you want/prefer  and i will adjust the code. :-)

pcuzner replied:
I think I wrote the test cases prior to adding the exclusive lock parameter. The code only uses excl for update calls - GET calls don't use it and don't need to lock.

The inventory object is not created upon each call. For a GET request the object doesn't lock, but when we said add_group - we create the inventory object and lock the underlying file to prevent other writers. the 'save' method is only invoked for update methods anyway.

I think the problem that we have is that the unittest needs to be restructured - perhaps to create the inventory object within each testcase to open it correctly

jmolmo replied:
Ok. I understand now.  I will fix the unit tests and restore the code to the previous state ( improving it if that is possible :-)

jmolmo commented:
Modifications:
-------------------
I have modified the unit tests in order to:
1. Unit test are now independent. Order of execution does not matter. (and i try to check only one thing in each test)
2. Test the current functionality included in the Ansible Inventory class
3. Address the Paul's suggestions 

I have find some things that i don't like very much:
- It seems that fcntl locks does not work .... so the mode "exclusive" is not really happening.
- With the current design each time a save operation is executed ( and this happen each time a modification over hosts/groups happen) the file descriptor of the inventary file is closed, so this implies that if you want to do more than  one operation you should use two AnsibleInventary Objects.

I would like more an operation mode where we will only lock the file when a write is being executed. Allowing  to keep one AnsibleInventory object alive and able to do any number of write operations.
As we are thinking what to do with the HA, maybe we can  avoid to use files for data with concurrency operations. 

I have added a destructor in the class, because we leave fd opened when use a exclusive mode object and no save operations were executed. (Unit test are good for detect this things.)

I would like to add a report of coverage to the project. I have seen that Tavis and Codecov works well together... but if you prefer another tool , for me is OK. Maybe create a task for this will be interesting.

Another thing i have seen,  maybe  we should move logs to systemD ( journalctl) , if we provide a SystemD service file ...it does not have too much sense to use log files.

pcuzner commented:
fcntl does work - but it doesn't block an open, only another lock request
```
------------------------------------ SESSION 1 -------------------------------------------------------------------------
[paul@rh460p Downloads]$ python3
Python 3.6.6 (default, Jul 19 2018, 14:25:17)
[GCC 8.1.1 20180712 (Red Hat 8.1.1-5)] on linux
Type "help", "copyright", "credits" or "license" for more information.         
>>> import fcntl
>>>
>>> fd = open("testfile", "r+")
>>> fcntl.flock(fd, fcntl.LOCK_EX|fcntl.LOCK_NB)
>>> # i have the lock
...
>>> # releasing the lock
...
>>> fcntl.flock(fd, fcntl.LOCK_UN)
>>> 

------------------------------------ SESSION 2 -------------------------------------------------------------------------
[paul@rh460p Downloads]$ python3
Python 3.6.6 (default, Jul 19 2018, 14:25:17)
[GCC 8.1.1 20180712 (Red Hat 8.1.1-5)] on linux
Type "help", "copyright", "credits" or "license" for more information.         
>>> import fcntl
>>> fd = open("testfile", "r+")
>>> fcntl.flock(fd, fcntl.LOCK_EX|fcntl.LOCK_NB)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
BlockingIOError: [Errno 11] Resource temporarily unavailable                   
>>> # I can't get the lock
...
>>> # lock above release, now I try and acquire again                          
...
>>> fcntl.flock(fd, fcntl.LOCK_EX|fcntl.LOCK_NB)                               
>>> # now I have the lock
...
>>> 
```

We can talk about the other issues tomorrow - thanks for the test case updates!
